### ISSUE TITLE
Migrating maxCapacity from v1 to vi2

### ISSUE THREAD

**Issue by bojanth** [alitto/pond#80]
Hello,

Thank you for creating such a useful library! ⭐ 

I'm trying to migrate my code from v1 to v2, but I'm not sure how to achieve the same functionality.
In v1 I was able to provide `maxCapacity` parameter to `pond.New()` function that would block submission of new tasks until there is available worker.
In v2 all tasks are submitted without blocking.

My use-case:
I have a lot of tasks that I try to execute in limited time. When the time expires, I wait for the submitted tasks to finish and ignore the remaining tasks.

Basically:
```go
	for _, task := range tasks {
		select {
		case <-ticker.C:
			e.logger.Info("the maximum allowed time expired, stopping the execution")
			break
		default:
			e.pool.Submit(func() {
				...
			})
		}
	}

	e.StopAndWait()
```

What would be a good way to implement this with v2?

**Comment by alitto** [comment]
Hey @bojanth!

Thanks for sharing your use-case, this is really helpful :slightly_smiling_face: 

As you point out, in v2 there wasn't any obvious way to implement the use-case you shared. Given that all pools are non-blocking by default, there isn't yet a way to block execution during task submission (`Submit()`).
I'm still wondering if we should provide an option to build a blocking pool (in terms of v1, that would be the equivalent to `maxCapacity=0`).

However, I've just shipped [v2.1.0](https://github.com/alitto/pond/releases/tag/v2.1.0), which includes some new features that could be useful to implement this particular use-case, although in a different style. Here's an example:

```go
pool := pond.NewPool(10)

// Create a context with a timeout of 5 seconds
timeout, cancel := context.WithTimeout(context.Background(), 5*time.Second)
defer cancel()

// Create a group with the timeout context
group := pool.NewGroupContext(timeout)

// Submit all tasks to the group 
for _, task := range tasks {
    group.Submit(task)
}

// Wait for them to complete or the timeout to expire
err := group.Wait()

if err != nil {
    fmt.Printf("Group completed with error: %v\n", err)
} else {
    fmt.Println("Group completed successfully")
}
```
And since `group.Submit` supports passing a variable number of tasks, you could replace the for loop with a single line:
```go
// Submit all tasks to the group and wait for them to complete or the timeout to expire
err := group.Submit(tasks...).Wait()
```

This approach is different because we are always submitting all tasks to the pool, but I think it has some benefits:
- Simplified program flow (no need to create a `select` statement)
- The `Wait()` method also cover the scenario where the pool's context is canceled.
- You get an explicit error that you could use to determine wether all tasks completed (`err == nil`) or the timeout expired (`err == context.DeadlineExceeded`)

Please let me know your thoughts on this.

Cheers!

**Comment by bojanth** [comment]
Hi @alitto,

Thank you for the quick and detailed response, I really appreciate it!
Your approach will definitely work for many situations. The `NewGroupContext()` is a cool feature! 🚀 

However, my current use-case is a bit special. I have to wait for every task that started to finish, without stopping it in the middle.

To give you a bit more context.
Every task does two things, first it uploads a video stream (this takes around 2 min), and after that it uploads a metadata file. If I would use a cancel context, it would stop the task somewhere in the middle, with partially uploaded video or missing metadata.

I've experimented with this hacky solution, which is not completely correct, but would probably be good enough in my case:
```go
package main

import (
	"fmt"
	"github.com/alitto/pond/v2"
	"time"
)

func main() {
	var concurrency int64 = 2
	pool := pond.NewPool(int(concurrency))
	ticker := time.NewTicker(time.Second * 8)

	tasks := []string{"a", "b", "c", "d", "e", "f", "g", "h", "i"}
	i := 0
out:
	for i < len(tasks) {
		select {
		case <-ticker.C:
			fmt.Println("the maximum allowed time expired, stopping the execution")
			break out
		default:
			running := pool.RunningWorkers()
			waiting := pool.WaitingTasks()
			// The following check is problematic.
			// Since tasks are added in an async way,
			// there is a chance that multiple tasks will be submitted before running or waiting is updated.
			// This issue could be minimized with a short sleep after the submission.
			if running < concurrency && waiting == 0 {
				fmt.Printf("waiting: %d, running: %d\n", waiting, running)
				// Pin i to avoid race conditions.
				j := i
				fmt.Printf("adding task %s\n", tasks[j])
				pool.Submit(func() {
					task(tasks[j])
				})
				i++
			}
		}
	}

	pool.StopAndWait()
}

func task(name string) {
	time.Sleep(time.Second * 5)
	fmt.Println("task done ", name)
}
```

I will think a bit more about this. There is probably a better solution that I don't see.

Thank you for the help! I will leave it up to you to decide if this issue should be closed.

**Comment by alitto** [comment]
I see, thanks for the extra details :+1: it helps to understand the use-case a bit more.
Just to clarify, the `timeout` context passed to the group in `pool.NewGroupContext(ctx)` will not interrupt running tasks upon cancellation, it only causes the pool to ignore tasks that haven't yet been started (queued tasks). So it shouldn't interrupt uploads that are in progress unless you explicitly pass the same context to the request, e.g. via ([`http.NewRequestWithContext`](https://github.com/alitto/pond/blob/main/examples/task_group/main.go#L14)).
 
An extra tip, you might also want to use the new `pond.ResultPool` if you need to collect some result after uploading the video and metadata. Here's an example of how it would look like:
 ```go
func main() {
    pool := pond.NewResultPool[UploadResult](10)

    timeout, cancel := context.WithTimeout(context.Background(), 60*time.Second)
    defer cancel()

    // Create a group with the timeout context
    group := pool.NewGroupContext(timeout)

    // Submit all tasks to the group 
    for _, task := range tasks {
        group.SubmitErr(func() (UploadResult, error) {
            return uploadVideo(task)
        })
    }

    // Wait for all of them to complete, the timeout to expire or the first error to occur, whichever comes first.
    results, err := group.Wait()

   if err != nil {
   
}

func uploadVideo(input string) (UploadResult, error) {
    // Upload video and metadata here
}

 // A struct with some details about the upload
 type UploadResult {
   ID string
   UploadedBytes uint
 }

```

**Comment by bojanth** [comment]
You are right, the timeout is not canceling ongoing tasks, it was my program that was exiting after the `group.Wait()`.
I assumed the started tasks will be finished.
If I call `pool.StopAndWait()` instead, everything works as expected.

It's a bit confusing that there is difference in behavior between these two functions. But I understand that not waiting for tasks to finish in the case of a timeout makes sense as well.
Maybe the best would be to support both cases? Via a separate method or by allowing options to `Wait()` method (e.g. `WaitOption.ImmediateTimeout`, `WaitOption.CompleteRunningTasks`).

**Comment by alitto** [comment]
Yes, I agree the semantics of the `Wait()` method were a bit confusing in this case.
I made some changes and released them as part of [v2.1.1](https://github.com/alitto/pond/releases/tag/v2.1.1) to change this and ensure the call to `group.Wait()` waits for ongoing tasks to complete before returning. I think this makes the behaviour more intuitive.
If there is a use case where one doesn't need to wait for running tasks to complete that can also be modeled by passing the same `ctx` to each task directly.

If you can give this new version a try that would be great. The same code as before should now behave as intended (the call to `group.Wait()` does wait for ongoing uploads to complete before returning.

Thanks again for your feedback!

**Comment by bojanth** [comment]
Awesome, thank you! I've tried `v2.1.1` and it works as expected. 🎉
